azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -318,10 +326,10 @@ def name() -> str:
         return "UserPassTokenProvider"
 
     def context(self) -> dict:
-        return {"authority": self._auth, "client_id": self._cloud_info.kusto_client_app_id, "username": self._user}
+        return {"authority": self._cloud_info.authority_uri(self._auth), "client_id": self._cloud_info.kusto_client_app_id, "username": self._user}

```

Comment:
e.g. this could be an issue if context is called after a failure do to endpoint

Reply:
You are right, flows like this could happen, and that's a big problem.

I see two options:
1. Add a call to init_cloud in the beginning of relevant context method.
The problem with that is the same as calling init_cloud in __init__ - it may unexpectedly block and we have no async equivalent
2.  Do a "best effort" - add a method to retrieve the cloud info, but only if it's already cached. Otherwise return None or the default cloud. This isn't great either since the users might get wrong/incomplete data.

I don't like either of these,  I'm not sure what's the correct way here.

Reply:
(ATM) Context is called within auth requests to add, well... context to error exceptions.
I think (1) would be fine

Reply:
Is this ok?